packages/exhibitor/extra/dcos_zk_backup.py

```diff
@@ -0,0 +1,350 @@
+    tmp_zookeeper_dir = tmp_dir / 'zookeeper'
+
+    print('Backing up ZooKeeper into {backup} via {tmp_zookeeper_dir}'.format(
+        backup=backup,
+        tmp_zookeeper_dir=tmp_zookeeper_dir,
+    ))
+
+    print('Validate that ZooKeeper is not running')
+    assert not _is_zookeeper_running(verbose)
+
+    print('Copying ZooKeeper files to {tmp_zookeeper_dir}'.format(
+        tmp_zookeeper_dir=tmp_zookeeper_dir,
+    ))
+
+    _copy_dir_and_preserve_ownership(
+        src=zookeeper_dir,
+        dst=tmp_zookeeper_dir,
+        verbose=verbose,
+    )
+
+    print('Creating ZooKeeper backup tar archive at {backup}'.format(backup=backup))
+
+    def _tar_filter(tar_info: tarfile.TarInfo) -> Optional[tarfile.TarInfo]:
+        if 'myid' in tar_info.name:
+            return None
+        if 'zookeeper.out' in tar_info.name:
+            return None
+        return tar_info
+
+    with tarfile.open(name=backup, mode='x:gz') as tar:
+        tar.add(
+            name=str(tmp_zookeeper_dir),
+            arcname='./zookeeper',
+            filter=_tar_filter,
+        )
+        if verbose:
+            tar.list()
+
+    print('Deleting temporary files in {tmp_zookeeper_dir}'.format(
+        tmp_zookeeper_dir=tmp_zookeeper_dir,
+    ))
+    shutil.rmtree(path=str(tmp_zookeeper_dir), ignore_errors=True)
+
+    print('ZooKeeper backup taken successfully')
+
+
+def restore_zookeeper(backup: Path, tmp_dir: Path, verbose: bool) -> None:
+    """
+    DC/OS ZooKeeper restore from backup procedure.
+
+    ZooKeeper changes files while it is running.
+    In order to have a consistent restore, one must stop ZooKeeper on all
+    master nodes, execute this procedure on all master nodes, and then start
+    ZooKeeper again on all master nodes.
+
+    Stopping ZooKeeper on all master nodes inevitably causes the DC/OS cluster
+    to experience downtime.  The restore procedure is intended for recovering
+    from an unusable DC/OS cluster as a last resort measure.
+
+    Running this script requires at least as much free space as the ZooKeeper backup takes.
+    """
+    zookeeper_dir = EXHIBITOR_DIR / 'zookeeper'
+    tmp_zookeeper_dir = tmp_dir / 'zookeeper'
+
+    print('Restoring local ZooKeeper instance from {backup}'.format(backup=backup))
+
+    print('Validate that ZooKeeper is not running')
+    assert not _is_zookeeper_running(verbose)
+
+    print('Moving ZooKeeper files temporarily to {tmp_zookeeper_dir}'.format(
+        tmp_zookeeper_dir=tmp_zookeeper_dir,
+    ))
+    # We copy files rather than move them so that if this script is interrupted
+    # while running, no data is lost inplace.
+    # However, this has a downside that we may use a lot of disk space -
+    # we have seen ZooKeeper files up to 20 GB in size.
+    _copy_dir_and_preserve_ownership(
+        src=zookeeper_dir,
+        dst=tmp_zookeeper_dir,
+        verbose=verbose,
+    )
+    shutil.rmtree(path=str(zookeeper_dir), ignore_errors=True)
```

Comment:
Why do we want to ignore errors here? Wouldn't a failure to remove the directory tree be a problem?

Reply:
Good catch, that was left over from the `rmtree` below.